zilliqa/src/consensus.rs

```diff
@@ -155,7 +150,7 @@ pub struct Consensus {
     /// The persistence database
     db: Arc<Db>,
     /// Actions that act on newly created blocks
-    transaction_pool: TransactionPool,
+    transaction_pool: RefCell<TransactionPool>,
```

Comment:
`RefCell<TransactionPool>` seems like a bit of a smell to me. If two people call `get_pending_block` at the same time (which the `&self` signature suggests should be possible), the `borrow_mut` call will panic as both try to take exclusive access over the `transaction_pool`. Its probably true that this doesn't happen in practice now, since the API layer accesses `Consensus` through a `Mutex`, but relying on this fact for correctness feels fragile. Its probably not far in our future that we need to move the locking further down the call-stack to improve API performance, at which point this assumption would break and the compiler won't tell us about the impending panics.

I guess our fundamental problem is that to serve 'pending' blocks we need to mutate the transaction pool to construct that pending block. Theoretically I can see a few solutions:
* Clone the stuff we need to mutate (what we do in `main` currently)
* Use a `RefCell`/`Mutex` to check/enforce exclusive access to the pending block (what this PR does)
* Change a bunch of signatures from `&self` to `&mut self` to enforce exclusive access to the pending block.
* Change when we assemble the 'pending' block - e.g. if we filled the block with 'ready' transactions as soon as those transactions arrived, theoretically we wouldn't need to mutate anything when reading the pending block.
* Change `TransactionPool`'s data structures to let it return the list of ready transactions without a `&mut self`. I haven't thought exactly how we could do this, but there is nothing special about the current `BinaryHeap` it uses for transactions and I could imagine a sorted `Vec` working just as well?

Is my understanding correct?

Reply:
1. Cloning the stuff we need - yes, it would move the issue to the `get_pending_block` call e.g. when that end-point gets spammed. It is also why the current PR is written the odd way it is i.e. mutate and return/restore the txn pool i.e. trading memory with compute.

2. Yes.

3. It was the first thing that I tried, but the impact of this had a pretty big blast radius, touching lots of code, simply to allow the `get_pending_block` call to run. So, I decided to take a different route - inner mutation.

4. This is something that we could consider doing, but it would effectively require re-architecting the consensus process again.

5. The BinaryHeap uses an underlying Vec as-is. I could give this a try to see how well it works. Let me rethink this.